snapcraft/internal/pluginhandler/_runner.py

```diff
@@ -0,0 +1,193 @@
+                    call_fifo=call_fifo.path,
+                    feedback_fifo=feedback_fifo.path)))
+
+            process = subprocess.Popen(
+                ['/bin/sh', '-e', '-c', script], cwd=self._builddir)
+
+            status = None
+            try:
+                while status is None:
+                    function_call = call_fifo.read()
+                    if function_call:
+                        self._handle_builtin_function(
+                            scriptlet_name, function_call.strip())
+                        feedback_fifo.write('success\n')
+                    status = process.poll()
+            finally:
+                call_fifo.close()
+                feedback_fifo.close()
+
+            if status:
+                raise errors.ScriptletRunError(
+                    scriptlet_name=scriptlet_name, code=status)
+
+    def _handle_builtin_function(self, scriptlet_name, function_call):
+        function_call_components = shlex.split(function_call)
+        if len(function_call_components) == 0:
+            # This means a snapcraft developer messed up. Should never be
+            # encountered in real life.
+            raise ValueError(
+                '{!r} scriptlet somehow managed to call a function without '
+                'providing a function name: {}'.format(
+                    scriptlet_name, function_call))
+
+        if len(function_call_components) > 2:
+            # This also means a snapcraft developer messed up. Should never be
+            # encountered in real life.
+            raise ValueError(
+                '{!r} scriptlet called a function with too many args (a max '
+                'of 1 arg is supported): {}'.format(
+                    scriptlet_name, function_call))
+
+        function_name = function_call_components[0]
+        kwargs = {}
+        if len(function_call_components) == 2:
+            try:
+                kwargs = json.loads(function_call_components[1])
+            except json.decoder.JSONDecodeError as e:
+                # This also means a snapcraft developer messed up. Should
+                # never be encountered in real life.
+                raise ValueError(
+                    '{!r} scriptlet called a function with its arg as invalid '
+                    'json: {}'.format(scriptlet_name, function_call)) from e
+
+        try:
+            function = self._builtin_functions[function_name]
+            function(**kwargs)
+        except KeyError as e:
+            # This also means a snapcraft developer messed up. Should never
+            # be encountered in real life.
+            raise ValueError(
+                '{!r} scriptlet called an undefined builtin function: '
+                '{}'.format(scriptlet_name, function_call)) from e
+
+
+class _NonBlockingFifo:
+
+    def __init__(self, *, path: str, permissions) -> None:
+        os.mkfifo(path)
+        self.path = path
+        self._permissions = permissions
```

Comment:
I'd get rid of 'permissions' entirely, and always open the fifo `O_RDWR`. This'd allow you to drop the `_fifo_open` method entirely and just open in in `__init__`.

Reply:
Ah! A genius you are! That was the part of this that was really bothering me.

Reply:
merely a reader of manpages :-D